pkg/controller/controllerutil/controllerutil.go

```diff
@@ -32,6 +34,86 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 )
 
+var (
+	// Will store memoized controller namespace
+	inClusterNamespace     string
+	inClusterNamespacePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+	// ErrNotRunningInCluster is the error for when the controller is not running in cluster
+	ErrNotRunningInCluster = &NotRunningInClusterError{}
+	// ErrNamespaceReadError is the error when the controller cannot read its namespace
+	ErrNamespaceReadError = &NamespaceReadError{}
+)
+
+// ReadInClusterNamespace returns the namespace by which the controller is running under.
+// read from "/var/run/secrets/kubernetes.io/serviceaccount/namespace".
+func ReadInClusterNamespace() (string, error) {
+	data, err := os.ReadFile(inClusterNamespacePath)
+	if errors.Is(err, os.ErrNotExist) {
+		return "", &NotRunningInClusterError{cause: err}
+	} else if err != nil {
+		return "", &NamespaceReadError{cause: err}
+	}
+
+	return string(data), nil
+}
+
+// ReadInClusterNamespaceCached returns the namespace by which the controller is running under.
+// read from "/var/run/secrets/kubernetes.io/serviceaccount/namespace".
+// Note that the resolved namespace is cached after it is successfully read from disk.
+func ReadInClusterNamespaceCached() (string, error) {
+	if inClusterNamespace != "" {
```

Comment:
I think we should make this threadsafe

Reply:
Seemed a bit unnecessary given the very low impact, but have done so.